src/reactapp/src/components/paypalExpress/hooks/usePaypalExpress.js

```diff
@@ -0,0 +1,138 @@
+  const { selectedPaymentMethod, methodList } = usePaymentMethodCartContext();
+  const cartId = LocalStorage.getCartId();
+
+  /*
+   Check if is possible to proceed on placing the order.
+   */
+  useEffect(() => {
+    if (
+      query &&
+      selectedMethod?.methodCode &&
+      (selectedPaymentMethod?.code === '' ||
+        selectedPaymentMethod?.code === 'paypal_express')
+    )
+      setProcessPaymentEnable(true);
+  }, [query, setProcessPaymentEnable, selectedMethod, selectedPaymentMethod]);
+
+  useEffect(() => {
+    const processPaypalExpress = async () => {
+      setProcessPaymentEnable(false);
+      /*
+       Get paypal express required data from response query
+       */
+      const { token, PayerID: payerId } = getTokenPayerId(query);
+      const paymentMethod = _objToArray(methodList).find(
+        method => method.code === 'paypal_express'
+      );
+
+      if (
+        !token ||
+        !payerId ||
+        !cartId ||
+        !selectedMethod ||
+        !paymentMethod ||
+        !cartBillingAddress
+      ) {
+        return;
+      }
+
+      /*
+       Set payment method on cart and place the order
+       */
+      try {
+        setPageLoader(true);
+        await setPaymentMethodPaypalExpress({ payerId, token });
+        const response = await placeOrder();
+        if (response && response.order_number) {
+          LocalStorage.clearCheckoutStorage();
+          window.location.replace(
+            `${config.baseUrl}/checkout/onepage/success/`
+          );
+        }
+      } catch (error) {
+        setPageLoader(false);
+        setErrorMessage(
+          __(
+            'Something went wrong while adding the payment method to the quote.'
+          )
+        );
+      }
+    };
+    if (processPaymentEnable) {
+      processPaypalExpress(cartId, methodList);
+    }
+  }, [
+    cartId,
+    methodList,
+    processPaymentEnable,
+    query,
+    selectedMethod,
+    setErrorMessage,
+    setPageLoader,
+    cartBillingAddress,
+  ]);
+
+  /*
+   Get the customer token from the BE and redirect to paypal.
+   */
+  const authorizeUser = useCallback(async () => {
+    if (!selectedMethod?.carrierCode || !cartBillingAddress?.firstname) {
+      setErrorMessage(__('Please complete all the required data.'));
+      return;
+    }
+
+    setPageLoader(true);
+    const response = await createCustomerToken({
+      returnUrl: `checkout/index/index`,
+      cancelUrl: `checkout/index/index`,
+    });
+
+    if (response) {
+      const paypalExpressUrl =
+        response.createPaypalExpressToken?.paypal_urls.start;
+      if (!paypalExpressUrl) {
+        setErrorMessage('Paypal Error');
+        return;
+      }
+      window.location.href = paypalExpressUrl;
+    }
+  }, [selectedMethod, setErrorMessage, setPageLoader, cartBillingAddress]);
+
+  return {
+    authorizeUser,
+  };
+}
+
+/*
+ Utility to get the token and the payer id from the URL
+ */
+const getTokenPayerId = query => {
```

Comment:
No need for this. Use native `URLSearchParams` to get the parameters. see this link: https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams

With this, we can completely get rid off this helper function and retrieve token and payer id within the hook itself.

Reply:
Done

Reply:
Can we move `getTokenPlayerId` above the hook function? Since this helper function uses inside the hook, somone expect the definition of this helper function above the hook.